[Feat] Introduce device coordinator and enhance state management

MihaiSprinceana opened the pull request:
Closes https://github.com/edoeel/homeassistant-sabiana-hvac/issues/12

MihaiSprinceana committed:
Refactor Sabiana HVAC integration to introduce device coordinator and enhance state management

MihaiSprinceana committed:
Refactor Sabiana HVAC integration to introduce device coordinator and enhance state management

MihaiSprinceana commented:
Tested with success on my HA instance. 
PR is ready for code review.

edoeel commented:
Hi @MihaiSprinceana,

Thank you for your contribution, it's greatly appreciated! I'm currently going through a busy period, but I'll make time to review your Pull Request in the next few days. I'll also be happy to help with testing.

My initial thought is that calling the `getDevices` endpoint every 10 seconds might be too frequent. This could significantly increase traffic to the Sabiana servers and potentially lead to an account ban if they implement a rate limit.

Since the application already uses a socket, have you considered leveraging that for real-time updates instead of polling?

MihaiSprinceana commented:
@edoeel many thanks for the integration and for the suggestion. 
I've putted back the PR in draft mode to take some time to test the integration with the websocket. 
Opus 4.5 already created a fully functional websocket implementation, which I just tested on my HA, but I will test it more during the next days and review all the changes my self as well.

MihaiSprinceana committed:
Implement WebSocket support for real-time updates in Sabiana HVAC integration

MihaiSprinceana committed:
Implement WebSocket support for real-time updates in Sabiana HVAC integration

MihaiSprinceana committed:
Increase default poll interval to optimize WebSocket real-time updates

MihaiSprinceana committed:
Increase default poll interval to optimize WebSocket real-time updates

MihaiSprinceana committed:
Add mock fixtures for testing Sabiana HVAC integration and implement WebSocket manager tests

MihaiSprinceana committed:
Add mock fixtures for testing Sabiana HVAC integration and implement WebSocket manager tests

MihaiSprinceana commented:
@edoeel on my HA instance everything looks good. I will keep monitoring it during the next days and get back to you. If you have time please test it yourself and review the PR.

MihaiSprinceana committed:
code format

MihaiSprinceana committed:
code format

edoeel commented:
@MihaiSprinceana hey, I attached the video to better show you the problem. I reproduced the bug using the latest commit and pulling the updated Home Assistant images.

https://imgur.com/a/v5YDHO5

edoeel commented:
Hi @MihaiSprinceana! JTK are you still working on this PR?

MihaiSprinceana committed:
fix sleep mode decoding

MihaiSprinceana committed:
fix sleep mode decoding

MihaiSprinceana commented:
@edoeel I've fixed the Sleep mode decoding. Can you please have a look on your HA ? 
Thanks

AlessandroTischer commented:
@MihaiSprinceana hi! thanks for the contribution! Are you still working on it? Does it have any flaws?

MihaiSprinceana commented:
@edoeel @AlessandroTischer 
For me it's working quite good, with a problem which I need to investigate and fix : after a while the web socket connection drops and it needs a HA core restart. 

@edoeel I think that we might have diff devices, which send data in diff format. 

My HA along with sabiana app test:
https://imgur.com/a/2HZBiCt

MihaiSprinceana committed:
fix websocket reconnection

MihaiSprinceana committed:
fix websocket reconnection

MihaiSprinceana commented:
@edoeel when you have time can you please test my latest changes on your instance and let me know the output ? 

Thanks

MihaiSprinceana committed:
fix websocket reconnection

MihaiSprinceana committed:
fix websocket reconnection

scompaxxx commented:
Hi, first of all thanks for the work you've done with the integration! Just curious if you guys are still working on the polling of the real-time state.

edoeel commented:
@MihaiSprinceana I have performed some local testing comparing the current main branch with the changes in this PR. To conduct these tests simultaneously, I ran the main version on port 8124 and the PR version on port 8123.

Test Results:
1. Swing Mode: Horizontal/vertical swings controls have disappeared (present in main). **This is an issue**.
2. Dehumidifier: Button now correctly appears. **This is a feature**.
3. Websocket: Real-time updates are failing. I noticed a 2-minute delay, confirming the integration is relying on the getDeviceForUserV2 HTTP fallback instead of the websocket. **This is an issue**. No errors in logs, suggesting a silent failure.

<details><summary>Screenshots</summary>
<p>

<img width="1158" height="741" alt="8123 - 1" src="https://github.com/user-attachments/assets/4d07d394-91f3-478f-b366-eef7a1392248" />
<img width="1161" height="743" alt="8124 - 1" src="https://github.com/user-attachments/assets/31abf40d-197c-4a0e-8dbe-c0dcf0e5cc0d" />

</p>
</details>

edoeel committed:
fix websocket dns issue

edoeel commented:
@MihaiSprinceana I debugged the issue with real-time updates. I fixed in my local environment. You can find the fix [here](https://github.com/edoeel/homeassistant-sabiana-hvac/commit/04cf3e89f275c81bd050f92312aeda0235001db9).
<details><summary>Tech details of the fix</summary>
<p>

### Root cause (before fix)
The failure occurred before Socket.IO handshake/auth, inside the transport layer:
* `socketio.AsyncClient.connect()` -> `engineio` -> `aiohttp` DNS resolution.

**Runtime raised:**
* `TypeError: Channel.getaddrinfo() ...`
* `TypeError: Channel.query() ...`

The same resolver exception appeared in unrelated Home Assistant integrations (`radio_browser`, `homeassistant_alerts`), confirming an environment-level incompatibility in the async DNS path (`aiodns`/`pycares` API mismatch).

As a consequence, WebSocket never connected, so the integration operated only via REST fallback polling (`devices/getDeviceForUserV2`).

---

### Implemented fix
The WebSocket transport was made independent from the broken async DNS resolver by forcing a dedicated `aiohttp` session with a threaded resolver:

**In `websocket.py`, during connect:**
* Create `aiohttp` session:
  ```python
  aiohttp.ClientSession(connector=aiohttp.TCPConnector(resolver=aiohttp.ThreadedResolver()))
* Pass it to:
  ```python
  socketio.AsyncClient(..., http_session=self._http_session)

This bypasses the failing aiodns resolution path and uses thread-based/system DNS resolution for this Socket.IO client only.

Added robust session lifecycle handling:

* Close old HTTP session before rebuilding the client
* Close session on disconnect/shutdown

</p>
</details> 

The following points still need to be addressed:
- Websocket Fix: The suggested fix for the aiodns incompatibility (using ThreadedResolver) needs to be officially implemented in the PR to restore the transport layer.
- Swing Orientation: The regression regarding horizontal/vertical swing tilt controls is still present and needs a fix to match the main branch functionality.
- Temperature Control (Cooling Mode): I noticed an inconsistency with temperature updates in "Cool" mode:
  - App to HA: Changing the temperature in the official app does not update the state in Home Assistant.
  - HA to App: Changes made from Home Assistant are consistently reflected in the device/app.

Thanks for your help!

MihaiSprinceana committed:
Merge branch 'edoeel:main' into main

MihaiSprinceana committed:
fix swing orientation, cooling temp sync, and HVAC mode mapping

- Fix swing mode: restore swing_modes list and SWING_MODE feature flag
- Fix cooling temperature: use mode-aware setpoint selection (summer/winter/auto bytes)
- Fix current temperature: decode as 16-bit word instead of single byte
- Fix HVAC mode: map mode 2 to AUTO instead of DRY
- Add swing decoding from flap position and present-flag bytes
- Add 32 new decoder tests, fix climate test fixtures (157/157 passing)

MihaiSprinceana committed:
make AUTO hvac mode conditional based on device capability flag

Read autoModeAvailable (bit 2 of byte 7) from device data, matching
the Android app's behavior. Devices that don't support AUTO (e.g.
Carisma Fly) now correctly show only Cool, Heat, Fan Only modes.
Fix ruff lint issues: import sorting, line lengths, unused imports.

MihaiSprinceana committed:
refactor - fix and removed unused code

Fixed resource leak: refresh callback now properly unregistered on shutdown

MihaiSprinceana committed:
fix lint
